lib/async_storage.js

```diff
@@ -0,0 +1,103 @@
+'use strict';
+const { createHook, executionAsyncId } = require('async_hooks');
+const { WeakReference } = internalBinding('util');
+const { emitExperimentalWarning } = require('internal/util');
+
+emitExperimentalWarning('async_storage');
+// id => WeakReference<AsyncResource>
+const idResourcePairing = new Map();
+const storages = new Set();
+createHook({
+  init(asyncId, type, triggerAsyncId, resource) {
+    // save new resource
+    idResourcePairing.set(asyncId, new WeakReference(resource));
+    const triggerResourceWeak = idResourcePairing.get(triggerAsyncId);
+    if (triggerResourceWeak) { // it has a parent, let it be known
+      storages.forEach((storage) => {
+        storage._propagate(resource, triggerResourceWeak.get());
+      });
+    }
+  },
+  destroy(asyncId) {
+    idResourcePairing.delete(asyncId);
+  },
+  promiseResolve(asyncId) {
+    idResourcePairing.delete(asyncId);
+  }
+})
+  .enable();
+
+class AsyncStorage {
+
+  constructor() {
+    this.count = 0;
+    this.kstore = Symbol('async-storage');
+  }
+
+  /**
+   * propagate the context from a parent resource to a child one
+   */
+  _propagate(resource, triggerResource) {
+    const store = triggerResource[this.kstore];
+    if (store) {
+      resource[this.kstore] = store;
+    }
+  }
+
+  /**
+   * get the current store, i.e. the unique map associated with current
+   * async resource and (maybe) inherited from the parent async resource
+   * will return null if no map is currently associated
+   */
+  getStore() {
+    const asyncId = executionAsyncId();
+    const resourceWeak = idResourcePairing.get(asyncId);
+    if (resourceWeak === undefined) {
+      return ; // should this throw?
+    }
+    const resource = resourceWeak.get();
+    return resource[this.kstore];
+  }
+
+  /**
+   * creates a parent async resource and associates a context with it
+   * then the callback is called in that new async context
+   * at this point calling getStore will return a map
+   * if called within the callback chain of another start call a new map will
+   * replace the previous one.
+   */
+  enter(cb) {
+    storages.add(this);
+    this.count++;
+    return setImmediate(() => {
```

Comment:
Why is this asynchronous, and why is it using setImmediate instead of nextTick?

Reply:
Moving to nextTick, I am trying to avoid a situation I feel mistake prone where enter would be called more than one and would erase current context
```js
storage.enter() // -> new context
// some unrelated code
storage.enter() // -> another new context
```
however there is a somehow equivalent situation now too. 

If I get your comment we could have `enter` and `exit` synchronous and that would just associate a store to current asyncResource?

Reply:
I'd also prefer it to not be asynchronous if possible. Or if it has to be asynchronous, make it compatible with promises

Reply:
It seems that if I make it synchronous, I can't be sure I already hive the current resource available in `idResourcePairing`. I am looking for other solutions.

Reply:
I might have a hack there

Reply:
@mcollina @targos just did an update to make it synchronous but it is pretty hacky.